stdlib/public/core/UnsafeBufferPointer.swift.gyb

```diff
@@ -366,6 +389,140 @@ public struct Unsafe${Mutable}BufferPointer<Element>
+  }
+  
+  /// Assigns every element in this buffer's memory to a copy of the given value.
+  ///
+  /// The buffer’s memory must be initialized or the buffer's `Element`
+  /// must be a trivial type. 
+  ///
+  /// - Parameters:
+  ///   - repeatedValue: The instance to assign this buffer's memory to.
+  ///
+  /// Warning: All buffer elements must be initialized before calling this. 
+  /// Assigning to part of the buffer must be done using the `assign(repeating:count:)`` 
+  /// method on the buffer’s `baseAddress`. 
+  @_inlineable
+  public func assign(repeating repeatedValue: Element) {
+    guard let dstBase = _position else {
+      return
+    }
+
+    dstBase.assign(repeating: repeatedValue, count: _end! - dstBase)
+  }
+  
+% end
+
+  /// Executes the given closure while temporarily binding the memory referenced 
+  /// by this buffer to the given type.
+  ///
+  /// Use this method when you have a buffer of memory bound to one type and
+  /// you need to access that memory as a buffer of another type. Accessing
+  /// memory as type `T` requires that the memory be bound to that type. A
+  /// memory location may only be bound to one type at a time, so accessing
+  /// the same memory as an unrelated type without first rebinding the memory
+  /// is undefined.
+  ///
+  /// The entire region of memory referenced by this buffer must be initialized.
+  /// 
+  /// Because this buffer's memory is no longer bound to its `Element` type
+  /// while the `body` closure executes, do not access memory using the
+  /// original buffer from within `body`. Instead, use the `body` closure's
+  /// buffer argument to access the values in memory as instances of type
+  /// `T`.
+  ///
+  /// After executing `body`, this method rebinds memory back to the original
+  /// `Element` type.
+  ///
+  /// - Parameters:
+  ///   - type: The type to temporarily bind the memory referenced by this
+  ///     buffer. The type `T` must be layout compatible
+  ///     with the pointer's `Element` type. The type `T` does not have to be 
+  ///     the same size as the buffer’s `Element` type.
+  ///   - body: A closure that takes a ${Mutable.lower()} typed buffer to the
+  ///     same memory as this buffer, only bound to type `T`. The buffer argument 
+  ///     contains as many complete instances of `T` as can fit within the 
+  ///     region of memory referenced by the original buffer. The closure's
+  ///     pointer argument is valid only for the duration of the closure's
+  ///     execution. If `body` has a return value, that value is also used as
+  ///     the return value for the `withMemoryRebound(to:_:)` method.
+  /// - Returns: The return value, if any, of the `body` closure parameter.
+  ///
+  /// Warning: Unlike this method, the destination type `T` of 
+  /// `Unsafe${Mutable}Pointer.withMemoryRebound(to:capacity:_:)` requires `T` 
+  /// to be the same size as its `Pointee` type.
```

Comment:
yes it does. it has to do with how the type gets rebound by the builtins.

Reply:
I don't think we don't need a warning about the usage of an `UnsafePointer` method here in the documentation for an `UnsafeBufferPointer` method. The documentation for `UnsafePointer.withMemoryRebound` already covers this precondition, and it isn't relevant here. (If this is meant to be a code comment, then it's fine, but it should double-slashed and in the method body.)

Reply:
Maybe an even better option would be to relax the precondition in `UnsafePointer.withMemoryRebound` so that the two similar methods have the same semantics. cc @atrick

Reply:
The buffer version of `withMemoryRebound` has an additional constraint. `MemoryLayout<T>.stride == MemoryLayout<Element>.stride`. This is necessary because the user doesn't provide a count for the resulting number of `T` elements. This buffer API simply doesn't make sense without this precondition.

I think the warning reads wrong. It sounds the precondition is on the UnsafePointer API, but that's wrong.

Reply:
the warning is correct. The buffer pointer first calculates the number of bytes it contains, then floor divides that by `MemoryLayout<T>.stride`. Therefore, the number of bytes of memory bound to `T` is always less than the number of bytes that get rebound to `Element` after the closure returns and the method is safe no matter the stride of `T`. 

In `UnsafeMutablePointer`, `count * MemoryLayout<T>.stride` bytes get bound to `T` but only `count * MemoryLayout<Element>.stride` bytes get rebound to `Element`, so `MemoryLayout<T>.stride` must be less than or equal to `MemoryLayout<Element>.stride`. For simplicity, we force the two strides to be equal in this case.

Reply:
Oops, I just realized the new buffer version is actually dropping the constraint specified on UnsafePointer. Note that it's important to specify these constraints even though they aren't dynamically enforced because we need to be able to implement a sanitizer without breaking code that was once considered valid.

In short, the sane thing to do here is to have the same constraint on buffer as we do on pointer. (I actually didn't realize until now that the proposal was changing this condition.) Otherwise, we need to specify the expected behavior in all the other corner cases. That would be a lot of pain and complexity for a situation that no-one is interested in.

We can't simply relax the stride constraint for general types because T already needs to be layout compatible with Element. So for count > 1, their also stride needs to be identical.

Figuring out how to relax this in certain cases requires a discussion about special casing count==1, triviality, endianness, alignment, etc... Let's please not go there! Even if we could come up with a less strict constraint, developers wouldn't be able to understand it, so that's not useful.

Reply:
I thought the purpose of rebinding type was to insert sequence points between reads and writes on different types? or is there something else going on or planned to be going on. I always assumed the purpose of withMemoryRebound methods was to make sure 

1. all memory accesses as type `Pointee` on the region of memory in question occur before entering the closure or after returning from it

2. all memory accesses as type `T` on the region of memory in question occur after entering the closure and before returning from it

And the purpose of `Builtin.bindMemory` is to serve as an entry point for a sanitizer to do the bookkeeping to make sure the region of memory in question in each step is still the same region. We don’t violate any of our type state assumptions in the buffer case because 

1. All of the memory referenced by the buffer pointer passed to the closure is bound to `T` 

2. All of the memory that is bound to `T` is returned to `Element` after the closure returns 

is this right?

Reply:
Yes, absolutely right!

Additionally `withMemoryRebound` is reinterpreting all of the `Element` values stored in the source buffer as `T` values in the closure argument. It's semantically doing what `unsafeBitCast` does, but it's happening in-place for values that have memory locations. Note that `unsafeBitCast` has an enforced size constraint even though it's only operating on a single value.

Even though this is an unsafe API, we need to be able to specify the conditions under which the code is guaranteed to be memory safe (in the future, a sanitizer will hopefully verify that code meets those conditions). Memory safety is important to guarantee even if some variables may have unspecified values.

Layout compatibility is necessary to ensure memory safety--the reinterpreted values need to have heap references in the same locations. When it comes to a buffer of potentially multiple values, the only simple way to guarantee layout compatibility of subsequent values is to introduce a stride constraint.

You certainly could imagine someone wanting to rebind a buffer of homegenous tuples as a flattened buffer of tuple elements. But we would need to specify all the conditions under which that is safe and debate this on swift-evolution. That's not a problem we need or want to solve now, but it's easy to loosen restrictions later or add new APIs to specifically handle those cases.

...and we didn't even get into specifying what the implementation is actually supposed to do when the sizes or strides are mismatched, which is really unclear to me. In fact, we're absolutely relying on withMemoryRebound not to introduce any overhead at all for extra checks and calculations, otherwise we would have enforced the constraints dynamically in debug mode at least.

In short, keep strong restrictions in place until they become too restrictive.

Reply:
i don’t really care about this use case enough to hold up the proposal over it, but what if you want to fuse a buffer of `UInt8`s into `UInt32`s or something?

    let integers:[UInt32] = bytes.withMemoryRebound(to: UInt32.self) 
    {
        return $0.map(UInt32.init(bigEndian:))
    }

My PNG library does this a lot but with raw buffer pointers as a workaround. js. if you still disagree i’ll just take it out. It’ll save us an integer division at least

Reply:
That's a great use case for raw buffers!

Well, your implementation is memory safe for trivial values... However, even if we came up with a set of rules that allowed withMemoryRebound to reinterpret memory as differently sized  integers (which is probably not a good idea), then in practice the integer division alone is a serious sticking point here.

Reply:
You’re probably right tbh. Using raw buffers is a bit more awkward for things like matrix buffers where you have a bunch of matrices stored back to back as an array of `Float`s. But that’s really a job for fixed sized arrays, not buffer pointer rebinds.

Reply:
someone run the ci pls

Reply:
@atrick you have to do it twice because of the sha